File: src/structure/createGeneratedFile.ts
import {Configuration} from "../configuration/configuration.js";
import {generateImports} from "./generateImports.js";
import {createPackageName} from "./package/createPackageName.js";
import {packageToOutputFileName} from "./package/packageToFileName.js";
import {removeUnusedImports} from "./import/removeUnusedImports.js";

export function createGeneratedFile(
    packageChunks: string[],
    fileName: string,
    imports: string[],
    body: string,
    configuration: Configuration,
) {
    const packageName = createPackageName(packageChunks)

    const outputFileName = packageToOutputFileName(packageChunks, fileName, configuration)

    const resultImports = removeUnusedImports(imports, body)
        .concat(generateImports(outputFileName, configuration))
        .filter(Boolean)
        .join("\n")

    const disclaimer = configuration.disclaimer

    const content = [
        disclaimer,
        `package ${packageName}`,
        resultImports,
        body,
    ]
        .filter(Boolean)
        .join("\n\n")

    return content.trim() + "\n"
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/structure/createGeneratedFile.kt
package io.github.sgrishchenko.karakum.structure

import io.github.sgrishchenko.karakum.configuration.Configuration
import io.github.sgrishchenko.karakum.structure.import.removeUnusedImports
import io.github.sgrishchenko.karakum.structure.`package`.createPackageName
import io.github.sgrishchenko.karakum.structure.`package`.packageToOutputFileName
import js.array.ReadonlyArray

fun createGeneratedFile(
    packageChunks: ReadonlyArray<String>,
    fileName: String,
    imports: ReadonlyArray<String>,
    body: String,
    configuration: Configuration,
): String {
    val packageName = createPackageName(packageChunks)

    val outputFileName = packageToOutputFileName(packageChunks, fileName, configuration)

    val resultImports = (
            removeUnusedImports(imports, body) +
                    generateImports(outputFileName, configuration)
            )
        .filter { it.isNotEmpty() }
        .joinToString(separator = "\n")

    val disclaimer = configuration.disclaimer

    val content = arrayOf(
        disclaimer,
        "package $packageName",
        resultImports,
        body,
    )
        .filter { it.isNotEmpty() }
        .joinToString(separator = "\n\n")

    return content.trim() + "\n"
}
